File: emergency-triage-assistant/backend/services/cacheService.js
/**
 * In-memory cache with 120 second TTL
 * Cache lookup must complete in <2ms
 */

const crypto = require('crypto');

class FastCache {
  constructor(ttlSeconds = 120) {
    this.cache = new Map();
    this.ttlSeconds = ttlSeconds;
  }

  _normalizeKey(question, patientId = null) {
    let normalized = question.toLowerCase().trim();
    if (patientId) {
      normalized = `${patientId}:${normalized}`;
    }
    // Hash for consistent key length
    return crypto.createHash('md5').update(normalized).digest('hex');
  }

  get(question, patientId = null) {
    const startTime = performance.now();
    const key = this._normalizeKey(question, patientId);

    if (this.cache.has(key)) {
      const entry = this.cache.get(key);
      const age = (Date.now() - entry.timestamp) / 1000;

      if (age < this.ttlSeconds) {
        const lookupTimeMs = performance.now() - startTime;
        console.log(`Cache HIT (age: ${age.toFixed(1)}s, lookup: ${lookupTimeMs.toFixed(3)}ms)`);
        return entry.data;
      } else {
        // Expired, remove
        this.cache.delete(key);
        console.log(`Cache EXPIRED (age: ${age.toFixed(1)}s)`);
      }
    }

    const lookupTimeMs = performance.now() - startTime;
    console.log(`Cache MISS (lookup: ${lookupTimeMs.toFixed(3)}ms)`);
    return null;
  }

  set(question, data, patientId = null) {
    const key = this._normalizeKey(question, patientId);
    this.cache.set(key, {
      data,
      timestamp: Date.now()
    });
    console.log(`Cache SET (total entries: ${this.cache.size})`);
  }

  clearExpired() {
    const now = Date.now();
    const expiredKeys = [];

    for (const [key, entry] of this.cache.entries()) {
      const age = (now - entry.timestamp) / 1000;
      if (age >= this.ttlSeconds) {
        expiredKeys.push(key);
      }
    }

    expiredKeys.forEach(key => this.cache.delete(key));

    if (expiredKeys.length > 0) {
      console.log(`Cleared ${expiredKeys.length} expired cache entries`);
    }
  }

  clearAll() {
    this.cache.clear();
    console.log('Cache cleared');
  }

  stats() {
    return {
      total_entries: this.cache.size,
      ttl_seconds: this.ttlSeconds
    };
  }
}

// Singleton instance
const cache = new FastCache(120);

// Periodic cleanup every 60 seconds
setInterval(() => {
  cache.clearExpired();
}, 60000);

module.exports = { cache };

File: emergency-triage-assistant/fastapi-backend/app/services/cache_service.py
"""
In-memory cache with 120 second TTL
Cache lookup must complete in <2ms
"""
import time
from typing import Optional, Dict, Any
import hashlib
import logging

logger = logging.getLogger(__name__)

class FastCache:
    def __init__(self, ttl_seconds: int = 120):
        self.cache: Dict[str, Dict[str, Any]] = {}
        self.ttl_seconds = ttl_seconds
    
    def _normalize_key(self, question: str, patient_id: Optional[str] = None) -> str:
        """Create normalized cache key"""
        normalized = question.lower().strip()
        if patient_id:
            normalized = f"{patient_id}:{normalized}"
        # Hash for consistent key length
        return hashlib.md5(normalized.encode()).hexdigest()
    
    def get(self, question: str, patient_id: Optional[str] = None) -> Optional[Dict[str, Any]]:
        """
        Get cached response if exists and not expired
        Must complete in <2ms
        """
        start_time = time.perf_counter()
        
        key = self._normalize_key(question, patient_id)
        
        if key in self.cache:
            entry = self.cache[key]
            age = time.time() - entry['timestamp']
            
            if age < self.ttl_seconds:
                lookup_time_ms = (time.perf_counter() - start_time) * 1000
                logger.info(f"Cache HIT (age: {age:.1f}s, lookup: {lookup_time_ms:.3f}ms)")
                return entry['data']
            else:
                # Expired, remove
                del self.cache[key]
                logger.info(f"Cache EXPIRED (age: {age:.1f}s)")
        
        lookup_time_ms = (time.perf_counter() - start_time) * 1000
        logger.info(f"Cache MISS (lookup: {lookup_time_ms:.3f}ms)")
        return None
    
    def set(self, question: str, data: Dict[str, Any], patient_id: Optional[str] = None):
        """Store response in cache"""
        key = self._normalize_key(question, patient_id)
        self.cache[key] = {
            'data': data,
            'timestamp': time.time()
        }
        logger.info(f"Cache SET (total entries: {len(self.cache)})")
    
    def clear_expired(self):
        """Remove expired entries (run periodically)"""
        now = time.time()
        expired_keys = [
            key for key, entry in self.cache.items()
            if now - entry['timestamp'] >= self.ttl_seconds
        ]
        for key in expired_keys:
            del self.cache[key]
        
        if expired_keys:
            logger.info(f"Cleared {len(expired_keys)} expired cache entries")
    
    def clear_all(self):
        """Clear entire cache"""
        self.cache.clear()
        logger.info("Cache cleared")
    
    def stats(self) -> Dict[str, Any]:
        """Get cache statistics"""
        return {
            "total_entries": len(self.cache),
            "ttl_seconds": self.ttl_seconds
        }


# Singleton instance
cache = FastCache(ttl_seconds=120)
